### PR TITLE
Fixes #18779 - API organization_id can only be numeric

### PR DESCRIPTION
**Pull request by dLobatog**
In paths like 'api/v2/organizations/:organization_id/products', or
'api/v2/products?organization_id=:organization_id' when you try to use a
name (e.g "ACMECORP") as the organization id, Foreman returns a 404.

However, the organization_id can be a name in most Foreman API paths and
'katello/api/v2/organizations/:organization_id'.

The problem is that the API controller is only searching by ID, but it
should use the 'friendly' scope provided by 'friendly_id' to search by
both name, title (for nested orgs) or numeric id.

### CODE REVIEW

**Comment by ehelms** [comment]
When testing this, using the label didn't seem to work for me.

**Comment by dLobatog** [comment]
@ehelms Sorry - I didn't mean label, but *title*. How did you test it? Organization.friendly comes from https://github.com/theforeman/foreman/blob/develop/app/models/taxonomies/organization.rb#L3 . Candlepin labels will not be matched - that's expected

**Comment by ehelms** [comment]
Should the label field be matched? Outside of `id`, that is the one of the primary interfaces for identifying an organization.

**Comment by dLobatog** [comment]
@ehelms I don't think it's expected based on the API docs. Is there any API endpoint that expects Candlepin labels?

**Comment by ehelms** [comment]
Is name expected based on the API docs? I couldn't find it.

**Comment by dLobatog** [comment]
@ehelms About the docs, all `id` fields in https://theforeman.org/api/1.14/index.html are expected to match either the name or the numeric ID. 

Users usually have no way of knowing the ID other than going to the /api/v2/organizations index or checking the URL which is clunky. Also, it's rather strange that `katello/api/v2/organizations/ACMEORG` is supported, but `katello/api/v2/organizations/ACMEORG/products` or `katello/api/v2/organizations/ACMEORG/activation_keys` is not (they only work with numeric IDs before this PR).

File: app/controllers/katello/api/v2/api_controller.rb [hunk 1]
```diff
@@ -126,7 +126,7 @@ def sort_params
 
     def find_optional_organization
       org_id = organization_id
-      return if org_id.nil?
+      return if org_id.blank?
```
Comment (dLobatog) [inline app/controllers/katello/api/v2/api_controller.rb]: This also prevents even trying to find an organization with ID `""` (empty string)

**Comment by dLobatog** [comment]
Just realized a good deal of the organization finding code is exactly the same as the code in Foreman. I'll check another 2 or 3 controllers  that have their own org finding methods, but I believe it'd be better to reuse the stuff in Foreman API controllers. Don't merge yet please.

**Comment by dLobatog** [comment]
Actually it looks like most of the code for finding organizations is redundant in Katello given that Foreman already has that functionality in the base controller. I started using the methods from core and removed the find_organization methods from here.

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@ class Api::V2::ActivationKeysController < Api::V2::ApiController
 
     before_action :verify_presence_of_organization_or_environment, :only => [:index]
     before_action :find_environment, :only => [:index, :create, :update]
-    before_action :find_optional_organization, :only => [:index, :create, :show]
+    before_action :find_optional_nested_object, :only => [:index, :create, :show]
```
Comment (ehelms) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: What is a nested object? I only ask cause this goes from obvious what it is doing to largely unreadable. I realize this is trying to re-use resources from Foreman core but I'd ask that we keep the obvious naming and wrap this call to provide a better interface.

Comment (dLobatog) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: I'll wrap it and fix tests. According to the failures, I think a lot of of the permissions were not properly enforced before using the methods from core.

Comment (dLobatog) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: As for nested_resource, it's the first resource when you have 2 in the API. e.g:

`katello/api/v2/orgnaizations/:organization_id/products/:id` 

Product is the resource, Organization is the nested resource. Certainly if it requires an explanation it means the name is not very readable...

**Comment by dLobatog** [comment]
Let's see if all tests pass now - I haven't wrapped the calls to be more descriptive yet.

**Comment by ehelms** [comment]
[test]

**Comment by beav** [comment]
Is this PR still active? It has not been updated in 15 months.

**Comment by dLobatog** [comment]
@beav Not really, but this is still an issue in the latest Katello (just tested it). `katello/api/v2/organizations/myorganization/products` returns a 404, 'Organization with id myorganization not found', whereas 'api/v2/organizations/14/products` works. Pretty annoying

**Comment by johnpmitsch** [comment]
@dLobatog should we close this PR if its not going to be updated? We can try to get the issue on a sprint for someone to fix it

**Comment by jturel** [comment]
@dLobatog I'm going to close this PR since there hasn't been any activity on it. Please reopen when you can get around to rebasing, etc!

**Pull request closed without merging**